src/drivers/optical_flow/paw3902/PAW3902.cpp

```diff
@@ -633,68 +694,74 @@ PAW3902::RunImpl()
 		break;
 	}
 
-	// TODO: page 35 switching scheme
-
-	// As a minimum requirement, PAW3902JF should not operate with Shutter < 0x01F4 in Mode 2, and must switch to the next operation mode.
-
-	_flow_sum_x += delta_x_raw;
-	_flow_sum_y += delta_y_raw;
+	if (buf.data.SQUAL > 0) {
+		_flow_dt_sum_usec += dt_flow;
+		_flow_sum_x += delta_x_raw;
+		_flow_sum_y += delta_y_raw;
+		_flow_sample_counter++;
+		_flow_quality_sum += buf.data.SQUAL;
+
+	} else {
+		// reset
+		_flow_dt_sum_usec = 0;
+		_flow_sum_x = 0;
+		_flow_sum_y = 0;
+		_flow_sample_counter = 0;
+		_flow_quality_sum = 0;
+		return;
+	}
 
 	// returns if the collect time has not been reached
-	if (_flow_dt_sum_usec < _collect_time) {
-		perf_end(_sample_perf);
+	if (_flow_dt_sum_usec < COLLECT_TIME) {
 		return;
 	}
 
 	optical_flow_s report{};
 	report.timestamp = timestamp_sample;
+	//report.device_id = get_device_id();
 
-	report.pixel_flow_x_integral = (float)_flow_sum_x / 500.0f;	// proportional factor + convert from pixels to radians
-	report.pixel_flow_y_integral = (float)_flow_sum_y / 500.0f;	// proportional factor + convert from pixels to radians
+	float pixel_flow_x_integral = (float)_flow_sum_x / 500.0f;	// proportional factor + convert from pixels to radians
+	float pixel_flow_y_integral = (float)_flow_sum_y / 500.0f;	// proportional factor + convert from pixels to radians
 
-	// rotate measurements in yaw from sensor frame to body frame according to parameter SENS_FLOW_ROT
-	float zeroval = 0.0f;
-	rotate_3f(_yaw_rotation, report.pixel_flow_x_integral, report.pixel_flow_y_integral, zeroval);
+	// rotate measurements in yaw from sensor frame to body frame
+	const matrix::Vector3f pixel_flow_rotated = _rotation * matrix::Vector3f{pixel_flow_x_integral, pixel_flow_y_integral, 0.f};
+	report.pixel_flow_x_integral = pixel_flow_rotated(0);
+	report.pixel_flow_y_integral = pixel_flow_rotated(1);
 
-	report.frame_count_since_last_readout = _frame_count_since_last;
-	report.integration_timespan = _flow_dt_sum_usec;	// microseconds
+	report.frame_count_since_last_readout = _flow_sample_counter; // number of frames
+	report.integration_timespan = _flow_dt_sum_usec;              // microseconds
 
-	report.sensor_id = 0;
-	report.quality = buf.data.SQUAL;
+	report.quality = _flow_sample_counter > 0 ? _flow_quality_sum / _flow_sample_counter : 0;
 
-	/* No gyro on this board */
+	// No gyro on this board
 	report.gyro_x_rate_integral = NAN;
 	report.gyro_y_rate_integral = NAN;
 	report.gyro_z_rate_integral = NAN;
 
 	// set (conservative) specs according to datasheet
-	report.max_flow_rate = 5.0f;       // Datasheet: 7.4 rad/s
+	report.max_flow_rate = 7.4f;        // Datasheet: 7.4 rad/s
```

Comment:
I believe this was set more conservatively for the PMW3901 after real world testing.